### PR TITLE
feat: source api implement

### PR DESCRIPTION
**Pull request by narrowizard**
# Summary

### Key Points

- [ ] This is a breaking change
- [x] New or existing documentation is updated

### Description
#151 

### Does this close any open issues?
#151 

### Other Information
Now we can add plugin configuration via source api.

### CODE REVIEW

**Review by klesh — COMMENTED** [review]
LGTM

**Review by klesh — APPROVED** [review]

File: apps/rest/src/controllers/source.spec.ts [hunk 1]
```diff
@@ -34,34 +63,63 @@ describe('SourceController', () => {
       const reqListSource = {
         page: 1,
         pagesize: 10,
-        type: 'jira' as const,
+        type: 'jira',
       };
+
+      const fn = jest
+        .spyOn(sourceService, 'list')
+        .mockImplementation(async () => {
```
Comment (joncodo) [inline apps/rest/src/controllers/source.spec.ts:71]: I think this test needs to call 'expect' in order to be a real test

File: apps/rest/src/controllers/source.spec.ts [hunk 2]
```diff
@@ -34,34 +63,63 @@ describe('SourceController', () => {
...
 
   describe('get', () => {
     it('should return target source', async () => {
+      const fn = jest
+        .spyOn(sourceService, 'get')
+        .mockImplementation(async () => {
+          return new Source();
+        });
+
       await sourceController.get('id');
```
Comment (joncodo) [inline apps/rest/src/controllers/source.spec.ts:91]: I think this test needs to call 'expect' in order to be a real test

File: apps/rest/src/controllers/source.spec.ts [hunk 3]
```diff
@@ -34,34 +63,63 @@ describe('SourceController', () => {
...
           auth: 'base64EncodedAuthToken',
         },
       };
+      const fn = jest
+        .spyOn(sourceService, 'update')
+        .mockImplementation(async () => {
+          return new Source();
+        });
+
       await sourceController.update('id', reqUpdateSource);
```
Comment (joncodo) [inline apps/rest/src/controllers/source.spec.ts:111]: I think this test needs to call 'expect' in order to be a real test

File: apps/rest/src/controllers/source.spec.ts [hunk 4]
```diff
@@ -34,34 +63,63 @@ describe('SourceController', () => {
...
 
   describe('delete', () => {
     it('should delete source', async () => {
+      const fn = jest
+        .spyOn(sourceService, 'delete')
+        .mockImplementation(async () => {
+          return new Source();
+        });
+
       await sourceController.delete('id');
```
Comment (joncodo) [inline apps/rest/src/controllers/source.spec.ts:123]: I think this test needs to call 'expect' in order to be a real test

File: apps/rest/src/services/source.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    const sources = await this.em.getRepository(Source).find(options);
+    return {
+      offset,
+      total,
+      data: sources,
+    };
+  }
+
+  async create(data: CreateSource): Promise<Source> {
+    const source = new Source();
```
Comment (joncodo) [inline apps/rest/src/services/source.ts:35]: Can we just do `const source = new Source(data);`? That way, we don't have to manage the individual attributes here?

File: apps/rest/src/types/pagination.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ export class PaginationRequest {
 }
 
 export class PaginationResponse<T> {
-  page: number;
```
Comment (joncodo) [inline apps/rest/src/types/pagination.ts:15]: Don't we need page-size, offset and total to do proper paging?

File: apps/rest/test/source.e2e-spec.ts [hunk 7]
```diff
@@ -19,23 +24,33 @@ describe('SourceController (e2e)', () => {
...
       };
       return request(app.getHttpServer())
         .post('/source')
-        .send(sourceWithoutType)
-        .expect(400)
+        .send(newSource)
+        .expect(201)
         .expect((res) => {
-          expect(res.body).toMatchSnapshot();
+          expect(ignoreEntityProps(res.body)).toMatchSnapshot();
```
Comment (joncodo) [inline apps/rest/test/source.e2e-spec.ts:47]: How does this method know which snapshot you are referencing?

**Comment by joncodo** [comment]
It seems like sourceTask routes are not being loaded by Nest

Screen Shot 2021-08-20 at 9 48 04 AM

I created a source fine:

Screen Shot 2021-08-20 at 9 48 26 AM

But source task failed:

Screen Shot 2021-08-20 at 9 48 42 AM

I also see no tests for the sourceTask file

File: apps/rest/src/controllers/source.spec.ts (hunk 1 above)
Comment (narrowizard) [reply apps/rest/src/controllers/source.spec.ts:71]: You are right. In my previous plan, I will add some real test suit when `SourceController.list` contains its own logic.  
> `SourceController.list` only call `SourceService.list` method without any other codes now, maybe it make little sense to `expect` it.

File: apps/rest/test/source.e2e-spec.ts (hunk 7 above)
Comment (narrowizard) [reply apps/rest/test/source.e2e-spec.ts:47]: I think jest handled it well.
> jest combined `describe` content, `it` content and call order of `toMatchSnapShot` as a primary key, to match specified snapshot. (I guess)

File: apps/rest/src/services/source.ts (hunk 5 above)
Comment (narrowizard) [reply apps/rest/src/services/source.ts:35]: I think both solution is acceptable. No much ideas.

**Pull request merged**